x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -710,49 +711,171 @@ private static void writeMetadata(StreamOutput out, Map<String, Object> metadata
         }
     }
 
-    private void assertInternalConsistency() {
-        if (false == Assertions.ENABLED) {
-            return;
+    /**
+     * An Authentication object has internal constraint between its fields, e.g. if it is internal authentication,
+     * it must have an internal user. These logics are upheld when authentication is built as a result of successful
+     * authentication. Hence, this method mostly runs in test (where assertion is enabled).
+     * However, for RCS remote access, FC receives an authentication object as part of the request. There is
+     * no guarantee that this authentication object also maintains the internal logics. Therefore, this method
+     * is called explicitly in production when handling remote access requests.
+     */
+    public void checkConsistency() {
+        // isRunAs logic consistency
+        if (isRunAs()) {
+            assert authenticatingSubject != effectiveSubject : "isRunAs logic does not hold";
+        } else {
+            assert authenticatingSubject == effectiveSubject : "isRunAs logic does not hold";
+        }
+
+        // check consistency for each authentication type
+        if (isAuthenticatedAnonymously()) {
```

Comment:
Optional nit:

I think a switch case on the enum works quite cleanly here, and has the nice (admittedly small) benefit of having IDE support for auto-generating code for a new type if we ever add one:

```java
switch (getAuthenticationType()) {
    case ANONYMOUS -> checkConsistencyForAnonymousAuthenticationType();
    case INTERNAL -> checkConsistencyForInternalAuthenticationType();
    case API_KEY -> checkConsistencyForApiKeyAuthenticationType();
    case REALM -> checkConsistencyForRealmAuthenticationType();
    case TOKEN -> checkConsistencyForTokenAuthenticationType();
    default -> {
        assert false : "unknown authentication type " + type;
    }
}
```

Reply:
I updated with your suggestion. I think it is better to consistently use the enum check. Thanks!